### ISSUE TITLE
U.S. average generation profiles by fuel type

### ISSUE THREAD

**Issue by bl-young** [USEPA/ElectricityLCI#256]
We see a number of applications for providing fuel specific generation profiles at the national level, e.g., `Electricity - COAL - US`. At one point, I believe it was possible to create a version with aggregation to the entire U.S. (or at least I was able to do it with a few minor tweaks to the code). Do you anticipate still providing this option in the next release? The immediate applications we see for these mixes are as proxies to assess international grid mixes and to provide options for assessing changing U.S. grid mixes.

**Comment by bl-young** [comment]
FYI this is what I am aiming for: 
![image](https://github.com/user-attachments/assets/0594e45d-8898-499b-b6b6-d11cc95e6ceb)

(Note, I'm less concerned with the consumption mix and capturing potential imports from Canada). The changes I made in the commit I referenced above from a few years back seem to still work at the latest HEAD. I will do a PR for you to review.

Was there an alternate intention of `regional_aggregation: 'US'` if not what I'm showing in the screenshot?

**Comment by m-jamieson** [comment]
Nope, I believe the intent of regional_aggregation: "US" in aggregate_data in [generation.py](https://github.com/bl-young/ElectricityLCI/blob/8ce5b93b9280585766f01606cdb41a6983e52f5b/electricitylci/generation.py#L610) is as you show. However, I did not want to implement that in main.py, and if I recall __init__.py was made the way it is, so that it was easier to force BA-level trading and build up to the US consumption mix (and FERC).  The other benefit of doing it that way was being able to see the share of the US consumption mix from the various BAAs.

Anyhow, can try the pull request, but if it breaks the above, I'm inclined not to accept it. In that case, I think it would be worthwhile to make a new folder with some custom scripts to do this sort of thing.

**Comment by bl-young** [comment]
> The other benefit of doing it that way was being able to see the share of the US consumption mix from the various BAAs.

setting `regional_aggregation: "BA"` does that already, doesn't it? It creates a US consumption mix that shows the contribution from all the BAAs.

**Comment by dt-woods** [comment]
Fuel-specific generation rollups are different than a traded and distributed rollup. I think we can produce the generation inventory by fuel type.

**Comment by dt-woods** [comment]
## Proposal
Let $CM^{US}$ be the process, "Electricity; at grid; consumption mix - US - US", with an input table of several Balancing authority (BA) generation mix amounts, $gm^{BA}_i \forall i \in BA$, each of which point to a provider process of the same balancing authority, "Electricity; at grid; generation mix - BA" ($GM^{BA}$), such that:

$$
\sum_{CM^{US}\, \forall i \in BA} gm^{BA}_i = 1
$$ 

In other words, the mix amounts from each Balancing authority within the U.S. consumption mix sum to one.

Now, each $GM^{BA}$ process has an input table of several fuel category (FC) mix amounts, $lci^{BAi}_j \forall j \in FC$, such that: 

$$\sum_{GM^{BA}_i\, \forall j \in FC} lci^{BAi}_{j} = 1
$$

In other words, all fuel mixes within a Balancing authority's generation mix sum to one.

Therefore, I propose the following new processes: 

- The U.S. average generation mix, $GM^{US}$
- The U.S. average fuel-category inventories: $LCI^{US}_{FCj} \forall j \in FC$.

where:

$$
GM^{US} = lci^{US}_{FCj}\, \forall j \in FC
$$

and

$$
LCI^{US}_{FCj} = (gm^{BA}_i \times lci^{BAi}_{FCj})\, \forall i \in BA
$$

such that each $LCI^{US}\_{FCj}$ is a unit process providing the amount of $lci^{US}\_{FCj}$ in $GM^{US}$ for each fuel category, $j$.

**Comment by bl-young** [comment]
> Proposal

Yes I think that makes sense. That US consumption mix also will include some electricity inputs from imports right? (e.g. HydroQuebec). Not sure how that's going to impact things.

**Comment by dt-woods** [comment]
Presently, the international contributions from Canada have fuel category,
'All'. This is a part of the consumption mix, rather than the generation
mix. Not sure if these should be decoupled. Happy for suggestions.

**Comment by dt-woods** [comment]
Rather than use the mix amounts found in 'Electricity; at grid; consumption mix - US - US'; the generation mix may be accessed from `build_generation_data` in eia923_generation.py with a little help re-introducing the primary fuel categories (either through `eia923_primary_fuel` or appropriate method from generation_mix.py):

```python
# Basic setup
import logging
import electricitylci.model_config as config
from electricitylci.utils import get_logger

log = get_logger(True, False)
config.model_specs = config.build_model_class("ELCI_2022")

from electricitylci.eia923_generation import build_generation_data
from electricitylci.eia923_generation import eia923_primary_fuel
from electricitylci.generation_mix import create_generation_mix_process_df_from_model_generation_data

# METHOD 1:
# Reproduced from :func:`build_generation_data` by
# essentially re-adding the primary fuel info that was scraped.

# Read EIA 923 generation and primary fuel categories.
df = build_generation_data(generation_years=[2022,])
pf = eia923_primary_fuel(year=2022)
pf['Plant Id'] = pf['Plant Id'].astype('int')

# Merge electricity generation data and primary fuel categories
df = df.merge(pf, how='left', left_on='FacilityID', right_on='Plant Id')
df_fuel = df.groupby(by="FuelCategory").agg({
    'Electricity': 'sum', 'FacilityID': 'count'
})

# Calculate the generation mix of each fuel category
df_total = df['Electricity'].sum()
df_fuel['Mix'] = df_fuel['Electricity'] / df_total

# Look at the percents
df_fuel['Mix']*100

# METHOD 2:
# Uses the long-winded method from generation_mix.py; same result!
df = build_generation_data(generation_years=[2022,])
us = create_generation_mix_process_df_from_model_generation_data(df, 'US')
us = us.loc[us['FuelCategory'] != 'ALL', :]
us.set_index("FuelCategory")["Generation_Ratio"]*100
```

Results are the same in methods 1 and 2 above:

```
FuelCategory
BIOMASS          0.433854
COAL            17.105494
GAS             35.065594
GEOTHERMAL       0.403270
HYDRO            6.435533
MIXED            6.666364
NUCLEAR         19.187820
OIL              0.156038
OTHF             0.000058
SOLAR            3.519274
SOLARTHERMAL     0.075778
WIND            10.950924
```

Note the differences when compared to the consumption mix for the same year:

<img width="152" height="211" alt="Image" src="https://github.com/user-attachments/assets/72d0d8e8-e1da-4eda-aaa2-6906215789a5" />

Namely, the introduction of Canadian fuels (indicated by the 'ALL' category) and the reduction of a few select fuels (e.g., COAL, OIL, GAS, NUCLEAR) from net exports.

**Comment by dt-woods** [comment]
I noted that the above method has a few limitations, namely, it uses the full facility net generation and attributes it to the primary fuel category, which, upon reflection, the primary fuel often contributes <100% to the facility's net generation. Also, since only the primary fuel category at the facility level is used, mixed fuel types are ignored (e.g., a facility that burns coal and has solar). I opted to rewrite most of the methods found in eia923_generation.py, but keep everything at the reported fuel type code and prime mover, rather than at the facility level. After running the basic filters, keeping the state-level small generator aggregations (i.e., plant Id 99999), ignoring efficiency filter on OTHF fuel types (i.e., non-bio MSW, waste heat, tire-derived fuels), and omitting hydrogen fuels, I got this mix for U.S. average 2022:

```
FuelCategory
BIOMASS          0.583269
COAL            22.921555
GAS             30.387278
GEOTHERMAL       0.451827
HYDRO            7.147934
NUCLEAR         21.669854
OFSL             0.000547
OIL              0.521527
OTHF             0.114339
SOLAR            3.928206
SOLARTHERMAL     0.084240
WIND            12.189422
```

Notably, the MIXED category goes away because I examine each fuel type code in the EIA 923. Percentages are roughly the same. OIL jumps up, as it was likely found in the MIXED category.

The basic methods are as follows:

```python
from electricitylci.eia923_generation import eia923_download_extract
from electricitylci.eia923_generation import group_fuel_categories
from electricitylci.PhysicalQuantities import PhysicalQuantity as PQ

# NOTE: employ the same basic methods found in eia923_generation.py, but 
# at a granularity of prime mover + fuel, rather than facility + primary fuel.

# Get full listing of fuels by facility
eia923_gen_fuel = eia923_download_extract(year)

# Correct data type of Plant Id
eia923_gen_fuel['Plant Id'] = eia923_gen_fuel['Plant Id'].astype('int')

# Group by facility + prime mover + reported fuel type code
group_cols = [
    "Plant Id",
    "NAICS Code",
    "EIA Sector Number",
    "Reported Prime Mover",
    "Reported Fuel Type Code",
]
sum_cols = [
    "Net Generation (Megawatthours)",
    "Total Fuel Consumption MMBtu",
]
gen_df = eia923_gen_fuel.groupby(
    group_cols,
    as_index=False)[sum_cols].sum()

# Assign fuel categories to each fuel type code 
gen_df["FuelCategory"] = group_fuel_categories(gen_df)

# NOTE: the only missing fuel category in 2022 is 'HYDROGEN' (H2)

# Find solar thermal plants and correct their fuel category
criteria = (gen_df["Reported Fuel Type Code"] == "SUN") & (
    gen_df["Reported Prime Mover"] != "PV"
)
gen_df.loc[criteria, "FuelCategory"] = "SOLARTHERMAL"

# Apply the basic filters

# 1. Positive net generation 
if config.model_specs.include_only_egrid_facilities_with_positive_generation:
    f_crit = gen_df["Net Generation (Megawatthours)"] > 0
    gen_df = gen_df.loc[f_crit, :]

# 2. NAICS codes 
if config.model_specs.filter_non_egrid_emission_on_NAICS:
    # NOTE: Plant Id 99999 represents aggregated small generators that are not 
    # individually reported. Consider keeping in the U.S. average calculation.
    row_criteria = (
        (gen_df['NAICS Code'] == '99999') | (
            (gen_df['NAICS Code'] == '22') & (
                gen_df['EIA Sector Number'].isin(['1','2']))
        )
    )
    gen_df = gen_df.loc[row_criteria, :]

# 2. Efficiency
if config.model_specs.filter_on_efficiency:
    # Unit conversion for fuel consumption to energy equivalent
    btu = PQ("1 Btu")
    btu.convertToUnit("W*h")
    gen_df['efficiency'] = gen_df["Net Generation (Megawatthours)"] / (
        gen_df["Total Fuel Consumption MMBtu"] * btu.getValue()
    ) * 100

    upper = config.model_specs.egrid_facility_efficiency_filters[
        "upper_efficiency"
    ]
    lower = config.model_specs.egrid_facility_efficiency_filters[
        "lower_efficiency"
    ]
    # NEW: add 'OTHF' to exclude list; several of these have zero fuel
    # consumption because they're burning garbage or waste heat.
    fuel_categories_to_exclude = [
        "NUCLEAR", "WIND", "SOLAR", "SOLARTHERMAL", "HYDRO", "GEOTHERMAL",
        "OTHF"
    ]
    e_crit = (
        (gen_df["efficiency"] >= lower) & (
            gen_df["efficiency"] <= upper)
    ) | (
        gen_df["FuelCategory"].isin(fuel_categories_to_exclude)
    )
    gen_df = gen_df.loc[e_crit, :]

# Compute the fuel totals
df_fuel = gen_df.groupby(by="FuelCategory").agg({
    'Net Generation (Megawatthours)': 'sum', 
    'Plant Id': 'count'
})

fuel_tot = gen_df['Net Generation (Megawatthours)'].sum()
df_fuel['Mix'] = df_fuel['Net Generation (Megawatthours)']/fuel_tot

# Show the mix percentages
df_fuel['Mix']*100
```

**Comment by bl-young** [comment]
When I look at eGRID 2022 generation mix [here](https://www.epa.gov/system/files/documents/2024-01/egrid2022_summary_tables.pdf), it looks like their estimate of NG is quite a bit higher, nearly 40%. While each of coal, wind, and nuclear are a bit lower (~2% each). Any thoughts on the discrepancy there?

**Comment by dt-woods** [comment]
Well, using the "advanced" method proposed above loses the connection with the BA-level UPs, which include the MIXED category. I guess a third method where I use the simplified method, but incorporate the generation percent by primary fuel category, so as to not inflate the percentages.

**Comment by bl-young** [comment]
Are you expecting to validate the results of this generation mix with any existing piece of the U.S. baseline? I guess I'm asking how important it is to have this mix - that might dictate the approach. IMO the most important piece of this is getting the fuel specific processes and not necessarily the mix itself
